### ISSUE TITLE
Docs suggestion: useScrollRestoration

### ISSUE THREAD

**Issue by droganov** [ReactTraining/react-router#7452]
Scroll restoration example has a bug — scroll gets reset on page reload. Normally we expect it to survive.
Another issue is that we only want to reset on router push event.

Suggested react hook handles this situation:

```typescript
import { useHistory } from 'react-router-dom';
import { useEffect, useState } from 'react';

export function useScrollRestoration(): void {
  const history = useHistory();
  const [lastPathName, setLastPathName] = useState<string | null>(null);
  useEffect(() => {
    history.listen(({ pathname }, action) => {
      if (action === 'PUSH' && pathname !== lastPathName) window.scrollTo(0, 0);
      setLastPathName(pathname);
    });
  });
}
```

**Comment by timdorr** [comment]
The scroll restoration docs are really just a starting point. We're not looking to write and maintain a complete library for this. I would look to other libraries than using that code verbatim.

**Comment by fluidsonic** [comment]
+1

The docs should at least mention that the current examples it presents can mess up page reload and browser-based navigation. And that more work needs to be done manually - along with some clue where to look.
The pages randomly either scroll up or don't.
The documentation is misleading here and likely confuses developers.

I don't expect a generic solution for this. Just a helpful heads up.

**Comment by fluidsonic** [comment]
@droganov I think you example leaks a listener. And there's no need to maintain a state.

The following is working for me: Scroll on push, ignore back/forward/reload of browser.

I'm programming in Kotlin, but you get the idea.
```kt
react.component("ScrollToTop") {
    val history = useHistory()
    val location = useLocation()

    useEffect(history, location.pathname) {
        cleanup(history.listen { newLocation, action ->
            if (action == "PUSH" && newLocation.pathname != location.pathname)
                window.scrollTo(0.0, 0.0)
        })
    }
}
```

**Comment by droganov** [comment]
@fluidsonic location.pathname is not always last pushed location
for leaks... if you unmount the app you need to unsubscribe

**Comment by fluidsonic** [comment]
@droganov `useEffect` is called every time `useScrollRestoration` is called which may be on every render. Every time another listener is added.

@droganov the last pushed location should be irrelevant. Only the current location is relevant.
If I navigate to `a`, then push `b`, then back to `a`, then push `b` again it should scroll up, because we're coming from `a`.
OTOH that might again depend on the app and use case. Depending on the current location is closed to standard browser behavior without history API.

**Comment by droganov** [comment]
@fluidsonic maybe I agree with it, need to check
useEffect I guess is not called every time, it has no deps. Do you think it differs from `useEffect(effect, [])`?

**Comment by droganov** [comment]
@fluidsonic in your example `location.pathname` seem to be cached

**Comment by droganov** [comment]
@fluidsonic I checked your suggestion with the location hook, this code seems to be working for me:

```ts
export const ScrollRestoration = (): null => {
  const history = useHistory();
  const location = useLocation();
  useEffect(
    () =>
      history.listen(({ pathname }, action) => {
        if (action === "PUSH" && pathname !== location.pathname)
          window.scrollTo(0, 0);
      }),
    [location.pathname],
  );
  return null;
};
```

**Comment by fluidsonic** [comment]
> useEffect I guess is not called every time, it has no deps. Do you think it differs from `useEffect(effect, [])`?

`useEffect` without dependencies is executed on every render.
`useEffect` with `[]` dependencies is executed only once.

New example looks better 👍

**Comment by droganov** [comment]
@fluidsonic OK, then thank you for the update
